dot/parachain/dispute/queues.go

```diff
@@ -0,0 +1,230 @@
+
+	// PopBestEffort removes the next participation request from the best effort queue
+	PopBestEffort() *ParticipationItem
+
+	// PopPriority removes the next participation request from the priority queue
+	PopPriority() *ParticipationItem
+
+	// Len returns the number of items in the specified queue
+	Len(queueType ParticipationPriority) int
+}
+
+// QueueHandler implements Queue
+// It uses two btree's to store the requests. One for best effort and one for priority.
+// The queues store participationItem's.
+// The btree is ordered by the CandidateComparator of participationItem.
+type QueueHandler struct {
+	bestEffort *btree.BTree
+	priority   *btree.BTree
+
+	bestEffortLock sync.Mutex
+	priorityLock   sync.Mutex
+
+	bestEffortMaxSize int
+	priorityMaxSize   int
+
+	//TODO: add metrics
+}
+
+const (
+	// bestEffortQueueSize is the maximum size of the best effort queue
+	bestEffortQueueSize = 100
+	// priorityQueueSize is the maximum size of the priority queue
+	priorityQueueSize = 20000
+)
+
+func (q *QueueHandler) Queue(
+	comparator CandidateComparator,
+	request *ParticipationRequest,
+	priority ParticipationPriority,
+) error {
+	if priority.IsPriority() {
+		if q.priority.Len() >= q.priorityMaxSize {
+			return errorPriorityQueueFull
+		}
+
+		q.priorityLock.Lock()
+		q.priority.ReplaceOrInsert(newParticipationItem(comparator, request))
+		q.priorityLock.Unlock()
+	} else {
+		if q.bestEffort.Len() >= q.bestEffortMaxSize {
+			return errorBestEffortQueueFull
+		}
+
+		q.bestEffortLock.Lock()
+		q.bestEffort.ReplaceOrInsert(newParticipationItem(comparator, request))
+		q.bestEffortLock.Unlock()
+	}
+
+	return nil
+}
+
+func (q *QueueHandler) Dequeue() *ParticipationItem {
+	if item := q.PopPriority(); item != nil {
+		return item
+	}
+
+	return q.PopBestEffort()
+}
+
+func (q *QueueHandler) PrioritiseIfPresent(comparator CandidateComparator) error {
+	if q.priority.Len() >= priorityQueueSize {
+		return errorPriorityQueueFull
+	}
+
+	q.bestEffortLock.Lock()
+	// We remove the item from the best effort queue and add it to the priority queue if it exists
+	if item := q.bestEffort.Delete(newParticipationItem(comparator, nil)); item != nil {
+		q.priorityLock.Lock()
+		q.priority.ReplaceOrInsert(item)
+		q.priorityLock.Unlock()
+	}
+	q.bestEffortLock.Unlock()
+
+	return nil
+}
+
+func (q *QueueHandler) PopBestEffort() *ParticipationItem {
+	q.bestEffortLock.Lock()
+	defer q.bestEffortLock.Unlock()
+	if item := q.bestEffort.DeleteMin(); item != nil {
+		return item.(*ParticipationItem)
+	}
+
+	return nil
+}
+
+func (q *QueueHandler) PopPriority() *ParticipationItem {
+	q.priorityLock.Lock()
+	defer q.priorityLock.Unlock()
+	if item := q.priority.DeleteMin(); item != nil {
+		return item.(*ParticipationItem)
+	}
+
+	return nil
+}
+
+func (q *QueueHandler) Len(queueType ParticipationPriority) int {
+	if queueType.IsPriority() {
+		return q.priority.Len()
+	}
+
+	return q.bestEffort.Len()
+}
+
+var _ Queue = &QueueHandler{}
```

Comment:
avoid to allocate memory, ref:  https://stackoverflow.com/a/60663003/8157685
```suggestion
var _ Queue = (*QueueHandler)(nil)
```

Reply:
It's done, thank you!